File: apps/frameworks/model_downloader/cpp/src/model_sources.cpp
//
// Created by ruoyi.sjd on 2024/12/25.
// Copyright (c) 2024 Alibaba Group Holding Limited All rights reserved.
//

#include "model_sources.hpp"
#include <algorithm>

namespace mnn::downloader {

ModelSource ModelSources::FromString(const std::string& source_str) {
    std::string lower = source_str;
    std::transform(lower.begin(), lower.end(), lower.begin(), ::tolower);
    
    if (lower == "huggingface" || lower == "hf") return ModelSource::HUGGING_FACE;
    if (lower == "modelscope" || lower == "ms") return ModelSource::MODEL_SCOPE;
    if (lower == "modelers" || lower == "ml") return ModelSource::MODELERS;
    
    return ModelSource::UNKNOWN;
}

std::string ModelSources::GetDefaultOrgName(const std::string& provider) {
    std::string lower = provider;
    std::transform(lower.begin(), lower.end(), lower.begin(), ::tolower);
    
    if (lower == "huggingface" || lower == "hf") return "taobao-mnn";
    if (lower == "modelscope" || lower == "ms") return "MNN";
    if (lower == "modelers" || lower == "ml") return "MNN";
    
    return "MNN";  // Default to MNN for unknown providers
}

std::string ModelSources::ToString(ModelSource source) {
    switch (source) {
        case ModelSource::HUGGING_FACE: return SOURCE_HUGGING_FACE;
        case ModelSource::MODEL_SCOPE: return SOURCE_MODEL_SCOPE;
        case ModelSource::MODELERS: return SOURCE_MODELERS;
        default: return "Unknown";
    }
}

ModelSource ModelSources::GetSource(const std::string& model_id) {
    auto splits = SplitSource(model_id);
    return FromString(splits.first);
}

std::pair<std::string, std::string> ModelSources::SplitSource(const std::string& model_id) {
    size_t colon_pos = model_id.find('/');
    if (colon_pos == std::string::npos) {
        return {"", model_id};
    }
    
    std::string source = model_id.substr(0, colon_pos);
    std::string path = model_id.substr(colon_pos + 1);
    return {source, path};
}

std::string ModelSources::GetModelName(const std::string& model_id) {
    auto splits = SplitSource(model_id);
    if (splits.second.empty()) {
        return model_id;
    }
    
    // Extract the last part of the path as model name
    size_t last_slash = splits.second.find_last_of('/');
    if (last_slash != std::string::npos) {
        return splits.second.substr(last_slash + 1);
    }
    
    return splits.second;
}

} // namespace mnn::downloader

File: apps/frameworks/model_downloader/android/src/main/java/com/alibaba/mls/api/source/ModelSources.kt
// Created by ruoyi.sjd on 2025/2/11.
// Copyright (c) 2024 Alibaba Group Holding Limited All rights reserved.
package com.alibaba.mls.api.source

import com.alibaba.mls.api.download.R

class ModelSources {
    enum class ModelSourceType {
        MODEL_SCOPE,
        HUGGING_FACE,
        MODELERS,
        LOCAL
    }

    val remoteSourceType: ModelSourceType
        get() {
            return ModelSourceType.MODEL_SCOPE
        }

    private object InstanceHolder {
        val instance: ModelSources = ModelSources()
    }

    val config: ModelSourceConfig
        get() {
            if (mockConfig == null) {
                mockConfig = ModelSourceConfig.createMockSourceConfig()
            }
            return mockConfig!!
        }

    companion object {
        private var mockConfig: ModelSourceConfig? = null

        const val sourceHuffingFace = "HuggingFace"
        const val sourceModelers = "Modelers"
        const val sourceModelScope = "ModelScope"

        val sourceList = listOf(
            sourceHuffingFace,
            sourceModelers,
            sourceModelScope
        )

        val sourceDisPlayList = listOf(
            R.string.source_huggingface,
            R.string.source_modelers,
            R.string.source_modelscope
        )

        @JvmStatic
        fun get(): ModelSources {
            return InstanceHolder.instance
        }
    }
}
